spring-cloud-gcp-logging/src/main/java/org/springframework/cloud/gcp/logging/StackdriverJsonLayout.java

```diff
@@ -0,0 +1,229 @@
+		this.includeTraceId = includeTraceId;
+	}
+
+	/**
+	 * @return true if the spanId should be included into the JSON
+	 */
+	public boolean isIncludeSpanId() {
+		return this.includeSpanId;
+	}
+
+	/**
+	 * @param includeSpanId true if the spanId should be included into the JSON
+	 */
+	public void setIncludeSpanId(boolean includeSpanId) {
+		this.includeSpanId = includeSpanId;
+	}
+
+	/**
+	 *
+	 * @return true if the exception should be added to the message
+	 */
+	public boolean isIncludeExceptionInMessage() {
+		return this.includeExceptionInMessage;
+	}
+
+	/**
+	 * @param includeExceptionInMessage true if the exception should be added to the message
+	 */
+	public void setIncludeExceptionInMessage(boolean includeExceptionInMessage) {
+		this.includeExceptionInMessage = includeExceptionInMessage;
+	}
+
+	/**
+	 * @param event the logging event
+	 * @return the map which should get rendered as JSON
+	 */
+	@Override
+	protected Map<String, Object> toJsonMap(ILoggingEvent event) {
+
+		Map<String, Object> map = new LinkedHashMap<>();
+
+		if (this.includeMDC) {
+			event.getMDCPropertyMap().forEach((key, value) -> {
+				if (FILTERED_MDC_FIELDS.contains(key)) {
+					return;
+				}
+				map.put(key, value);
+			});
+		}
+		if (this.includeTimestamp) {
+			map.put(TIMESTAMP_SECONDS_ATTRIBUTE, TimeUnit.MILLISECONDS.toSeconds(event.getTimeStamp()));
+			map.put(TIMESTAMP_NANOS_ATTRIBUTE, TimeUnit.MILLISECONDS.toNanos(event.getTimeStamp() % 1_000));
+		}
+
+		add(SEVERITY_ATTRIBUTE, this.includeLevel, String.valueOf(event.getLevel()), map);
+		add(THREAD_ATTR_NAME, this.includeThreadName, event.getThreadName(), map);
+		add(LOGGER_ATTR_NAME, this.includeLoggerName, event.getLoggerName(), map);
+
+		if (this.includeFormattedMessage) {
+			String message = event.getFormattedMessage();
+			if (this.includeExceptionInMessage) {
+				IThrowableProxy throwableProxy = event.getThrowableProxy();
+				if (throwableProxy != null) {
+					String stackTrace = getThrowableProxyConverter().convert(event);
+					if (stackTrace != null && !stackTrace.equals("")) {
+						message += "\n" + stackTrace;
+					}
+				}
+			}
+			map.put(FORMATTED_MESSAGE_ATTR_NAME, message);
+		}
+		add(MESSAGE_ATTR_NAME, this.includeMessage, event.getMessage(), map);
+		add(CONTEXT_ATTR_NAME, this.includeContextName, event.getLoggerContextVO().getName(), map);
+		addThrowableInfo(EXCEPTION_ATTR_NAME, this.includeException, event, map);
+		addTraceId(event, map);
+		add(SPAN_ID_ATTRIBUTE, this.includeSpanId, event.getMDCPropertyMap().get(MDC_FIELD_SPAN_ID), map);
+		addCustomDataToJsonMap(map, event);
+		return map;
+	}
+
+	private void addTraceId(ILoggingEvent event, Map<String, Object> map) {
+		if (!this.includeTraceId) {
+			return;
+		}
+
+		String traceId = TraceLoggingEnhancer.getCurrentTraceId();
+		if (traceId == null) {
+			traceId = event.getMDCPropertyMap().get(MDC_FIELD_TRACE_ID);
+		}
+		if (!StringUtils.isEmpty(traceId) && !StringUtils.isEmpty(this.projectId)) {
```

Comment:
@Andy2003 The value of `this.projectId` seems to be `GOOGLE_CLOUD_PROJECT_IS_UNDEFINED`, when empty though. I think Logback might be translating from null to '...IS_UNDEFINED'? So, I don't think the condition is correct.

Reply:
You are right, it can also happen that the projectId will end with `_IS_UNDEFINED`. I added an additional check for this.